src/bloqade/emulate/ir/space.py

```diff
@@ -33,44 +33,57 @@ def create(cls, register: "Register"):
-            # defauly to 32 bit if smaller than 32 bit
-            config_type = np.result_type(min_int_type, np.uint32)
-            configurations = configurations.astype(config_type)
-
+            # default to 32 bit if smaller than 32 bit
+            configurations = np.arange(Ns, dtype=config_type)
             return Space(SpaceType.FullSpace, atom_type, sites, configurations)
 
-        for index_1, indices in enumerate(check_atoms):
-            # get which configurations are in rydberg state for the current index.
-            rydberg_configs_1 = atom_type.is_rydberg_at(configurations, index_1)
-            for index_2 in indices:  # loop over neighbors within blockade radius
-                # get which configus have the neighbor with a rydberg excitation
-                rydberg_configs_2 = atom_type.is_rydberg_at(configurations, index_2)
-                # get which states do not violate constraint
-                mask = np.logical_not(
-                    np.logical_and(rydberg_configs_1, rydberg_configs_2)
+        states = np.arange(atom_type.n_level, dtype=config_type)
+        configurations = states.copy()
+
+        for index_1, indices in enumerate(check_atoms, 1):
+            if len(indices) == 0:
+                continue
+
+            # loop over neighbors within blockade radius
+            # find all non-blockaded configurations
+            mask = np.logical_not(atom_type.is_rydberg_at(configurations, indices[0]))
+            for index_2 in indices[1:]:
+                is_not_rydberg = np.logical_not(
+                    atom_type.is_rydberg_at(configurations, index_2)
                 )
-                # remove states that violate blockade constraint
-                configurations = configurations[mask]
-                rydberg_configs_1 = rydberg_configs_1[mask]
+                np.logical_and(is_not_rydberg, mask, out=mask)
 
-        min_int_type = np.min_scalar_type(configurations[-1])
-        # defauly to 32 bit if smaller than 32 bit
-        config_type = np.result_type(min_int_type, np.uint32)
-        configurations = configurations.astype(config_type)
+            non_blockaded = configurations[mask]
+            np.logical_not(mask, out=mask)
+            blockaded = configurations[mask]
+
+            # add new configurations
+            # add all configurations because none of the configs are blockading
+            new_non_blockaded = np.kron(non_blockaded, np.ones_like(states)) + np.kron(
+                np.ones_like(non_blockaded), states * atom_type.n_level**index_1
+            )
+            # add all but the rydberg state because some at least one is blockading
+            new_blockaded = np.kron(blockaded, np.ones_like(states[:-1])) + np.kron(
+                np.ones_like(blockaded), states[:-1] * atom_type.n_level**index_1
+            )
+
+            configurations = np.hstack((new_blockaded, new_non_blockaded))
+
+        configurations.sort()
```

Comment:
it seems the only mutation happens here?

Reply:
I'll double check this.